readium/navigator/src/main/java/org/readium/r2/navigator/media2/MediaSessionNavigator.kt

```diff
@@ -0,0 +1,327 @@
+     */
+    suspend fun go(locator: Locator): Try<Unit, Exception> {
+        Timber.d("Go to locator $locator")
+        val itemIndex = checkNotNull(publication.readingOrder.indexOfFirstWithHref(locator.href))
+        val position = locator.locations.time ?: Duration.ZERO
+        return seek(itemIndex, position)
+    }
+
+    /**
+     * Seeks to the beginning of the given link.
+     */
+    suspend fun go(link: Link) =
+        go(link.toLocator())
+
+    /**
+     * Skips to a little amount of time later.
+     */
+    suspend fun goForward(): Try<Unit, Exception> =
+        seekBy(configuration.skipForwardInterval)
+
+    /**
+     * Skips to a little amount of time before.
+     */
+    suspend fun goBackward(): Try<Unit, Exception> =
+        seekBy(-configuration.skipBackwardInterval)
+
+    private suspend fun seekBy(offset: Duration): Try<Unit, Exception> =
+        this.playerFacade.playlist!!.durations
+            ?.let { smartSeekBy(offset, it) }
+            ?: dummySeekBy(offset)
+
+    private suspend fun smartSeekBy(
+        offset: Duration,
+        durations: List<Duration>
+    ): Try<Unit, Exception> {
+        val (newIndex, newPosition) =
+            SmartSeeker.dispatchSeek(
+                offset,
+                playerFacade.currentPosition!!,
+                playerFacade.currentIndex!!,
+                durations
+            )
+        Timber.d("Smart seeking by $offset resolved to item $newIndex position $newPosition")
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    private suspend fun dummySeekBy(offset: Duration): Try<Unit, Exception> {
+        val newIndex = playerFacade.currentIndex!!
+        val newPosition = playerFacade.currentPosition!! + offset
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    /**
+     * Stops the playback.
+     *
+     * Compared to [pause], the navigator may clear its state in whatever way is appropriate. For
+     * example, recovering a player's resources.
+     */
+    fun close() {
+        playerFacade.unregisterPlayerCallback(playerCallback)
+        playerCallback.close()
+        playerFacade.close()
+        coroutineScope.cancel()
+    }
+
+    fun session(context: Context, id: String, activityIntent: PendingIntent): MediaSession =
+        playerFacade.session(context, id, activityIntent)
+
+
+    data class Configuration(
+        val positionRefreshRate: Double = 2.0,  // Hz
+        val skipForwardInterval: Duration = 30.seconds,
+        val skipBackwardInterval: Duration = 30.seconds,
+    )
+
+    @ExperimentalTime
+    data class Playback(
+        val state: State,
+        val rate: Double,
+        val currentIndex: Int,
+        val currentLink: Link,
+        val currentPosition: Duration,
+        val bufferedPosition: Duration,
+    ) {
+
+        enum class State {
+            Playing,
+            Paused,
+            Finished,
+            Error
+        }
```

Comment:
`Idle` -> `Paused` is used.
`Loading` -> Presumably, the global state would be `Paused` or `Playing` which makes sense, I think.  Media2 has a special buffer state which I didn't explore. I'll have a look.

Reply:
Yeah for `Loading` in case of buffering that would be important to display a spinner in the user interface. Otherwise you get this feeling of something broken.

Reply:
I've added something for this.